File: solo5/bindings/hvt/gpio.c
#include "bindings.h"

void solo5_gpio_mode(uint8_t pin, uint8_t mode)
{
    volatile struct hvt_gpiomode md;
    md.pin = pin;
    md.mode = mode;
    md.ret = 0;
    
    hvt_do_hypercall(HVT_HYPERCALL_GPIOMODE, &md);

    assert(md.ret == 0);
}

void solo5_gpio_write(uint8_t pin, uint8_t value)
{
    volatile struct hvt_gpiowrite wr;
    wr.pin = pin;
    wr.value = value;
    wr.ret = 0;
    
    hvt_do_hypercall(HVT_HYPERCALL_GPIOWRITE, &wr);
    
    assert(wr.ret == 0);
}

uint8_t solo5_gpio_read(uint8_t pin)
{
    volatile struct hvt_gpioread rd;
    rd.pin = pin;
    rd.value = 0;
    rd.ret = 0;

    hvt_do_hypercall(HVT_HYPERCALL_GPIOREAD, &rd);

    assert(rd.ret == 0);
    return rd.value;
}

// vim: set sw=4 ts=4 sts=4 expandtab:

File: mirage-gpio-solo5/src/gpio.ml
(* I can't OCAML, so this is mostly just a bunch of cargo culting. If anything
 * seems really weird about this code, that's probably the reason. Feel free
 * to fix stuff if you know betther than me. *)

open Mirage_gpio
open OS.Solo5

(* Make the compiler STFU *)
type 'a io = 'a Lwt.t

type t = {
    name: string;
}

external solo5_gpio_mode:
    int -> int -> unit = "mirage_solo5_gpio_mode"
external solo5_gpio_write:
    int -> int -> unit = "mirage_solo5_gpio_write"
external solo5_gpio_read:
    int -> int = "mirage_solo5_gpio_read"

let disconnect _id =
    Lwt.return_unit

let connect name =
    Lwt.return ({ name })


let set_mode t pin mode =
    (* Make the compiler STFU *)
    let name = t.name in
    solo5_gpio_mode pin mode

let write t pin value =
    (* Make the compiler STFU *)
    let name = t.name in
    solo5_gpio_write pin value

let read t pin =
    (* Make the compiler STFU *)
    let name = t.name in
    solo5_gpio_read pin

(* vim: set sw=4 ts=4 sts=4 expandtab: *)
